File: ref/code/L05/perimeter.cpp
#include <bits/stdc++.h>
using namespace std;

struct point {
	int x, y;
};

bool icy[1001][1001], vis[1001][1001];
int dir[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};  // 方向数组
int n, areaMax = 0, p = INT_MAX;

void bfs(int i, int j) {
	queue<point> q;
	q.push({i, j});
    vis[i][j] = true;
    int area = 1, perimeter = 0;

    while(!q.empty()) {
        point cur = q.front();
        q.pop();
        for(int i=0; i<4;i++) { // 四个方向
        	//算出新的位置坐标
            int nx =  cur.x + dir[i][0];
            int ny =  cur.y + dir[i][1];
            //判断新的位置是否越界 或者 非冰激凌
            if(nx<1 || nx > n || ny < 1 || ny > n || !icy[nx][ny]) {
            	perimeter++;
            	continue;
			}

            // 已访问过 
            if(vis[nx][ny]) {
				continue;
			}

			q.push({nx, ny});
			vis[nx][ny] = true;
			area++;
		}
    }

    if(areaMax < area || areaMax == area && p > perimeter) {
    	areaMax = area;
    	p = perimeter;
	}
}

/*
6
##....
....#.
.#..#.
.#####
...###
....##

13 22
*/
int main() {
//	freopen("perimeter.in","r",stdin);
//	freopen("perimeter.out","w",stdout);
    ios::sync_with_stdio(false);
	cin.tie(0), cout.tie(0);
    cin>>n;
    
    for(int i=1; i<=n; i++) {
    	string s;
    	cin>>s;
    	for(int j=1; j<=n; j++) {
    		if(s[j-1] == '#') {
    			icy[i][j] = true;
			}
		}
	}
	
    for(int i=1; i<=n; i++) {
    	for(int j=1; j<=n; j++) {
    		if(icy[i][j] && !vis[i][j]) {
    			bfs(i, j);
			}
		}
	}

	cout<<areaMax<<" "<<p<<endl;

	return 0;
}

File: ref/code/L05/perimeter.java
package com.deltacoding.usaco.s2018_2019.jan2019.silver;

import java.io.*;
import java.util.*;

public class perimeter {
	static int n, maxArea=0, perimeter=Integer.MAX_VALUE; // 结果
	static boolean[][] icy;
	static boolean[][] vis;
	static int[][] DIR = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
	
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new FileReader("perimeter.in"));
		PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("perimeter.out")));
		n = Integer.parseInt(in.readLine());
		icy = new boolean[n][n];
		vis = new boolean[n][n];
		for(int i=0; i<n; i++) {
			String s = in.readLine();
			for(int j=0; j<s.length(); j++) {
				icy[i][j] = (s.charAt(j) == '#');
			}
		}
		
		for(int i=0; i<n; i++) {
			for(int j=0; j<n; j++) {
				if(icy[i][j] && !vis[i][j]) {
					bfs(i, j);
				}
			}
		}
		
		out.println(maxArea + " " + perimeter);
		in.close();
		out.close();
	}

	static void bfs(int i, int j) {
		Queue<Point> q = new ArrayDeque<>();
		int blobArea=1, blobPerimeter=0; // 面积 周长
		q.offer(new Point(i, j));
		vis[i][j] = true;
		while(!q.isEmpty()) {
			Point point = q.poll();
			for(int k=0; k<4; k++) {
				int nx = point.x + DIR[k][0];
				int ny = point.y + DIR[k][1];
				if(nx<0 || nx>=n || ny<0 || ny>=n || !icy[nx][ny]) {
					blobPerimeter++;
					continue;
				}
				if(vis[nx][ny]) {
					continue;
				}
				q.offer(new Point(nx, ny));
				vis[nx][ny] = true;
				blobArea++;
			}
		}

		if(blobArea > maxArea) {
			maxArea = blobArea;
			perimeter = blobPerimeter;
		} else if (blobArea == maxArea) {
			perimeter = Math.min(blobPerimeter, perimeter);
		}
	}
	
	static class Point {
		int x, y;
		public Point(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}
}